packages/@headlessui-react/src/components/transitions/transition.tsx

```diff
@@ -22,15 +22,20 @@ import {
 } from '../../utils/render'
 import { OpenClosedProvider, State, useOpenClosed } from '../../internal/open-closed'
 import { match } from '../../utils/match'
-import { microTask } from '../../utils/micro-task'
 import { useId } from '../../hooks/use-id'
-import { useIsMounted } from '../../hooks/use-is-mounted'
 import { useIsoMorphicEffect } from '../../hooks/use-iso-morphic-effect'
 import { useLatestValue } from '../../hooks/use-latest-value'
 import { useServerHandoffComplete } from '../../hooks/use-server-handoff-complete'
 import { useSyncRefs } from '../../hooks/use-sync-refs'
 import { useTransition } from '../../hooks/use-transition'
 import { useEvent } from '../../hooks/use-event'
+import { defer } from '../../utils/defer'
+
+type TransitionDirection = 'enter' | 'leave' | 'idle'
+let fakeDefer = () => defer<TransitionDirection>() // One day we will be able to do ReturnType<typeof genericFunction<T>>
```

Comment:
I thought this already worked?

```ts
async function genericFunction<T>(foo: T): Promise<T> {
  return foo
}

type Cool<T> = ReturnType<typeof genericFunction<T>>
type A = Cool<"foo">

// A === Promise<"foo">
```

Reply:
As does this:
```ts
type Cool2 = ReturnType<typeof genericFunction<"bar">> // Cool2 === Promise<"bar">
```

Reply:
Oh, my editor disagrees 😅 
<img width="2650" alt="image" src="https://user-images.githubusercontent.com/1834413/173443768-6234a8a4-4278-4f32-8187-dc582704c08d.png">

Reply:
The version we are on doesn't support this, but we can bump the version 👍

Reply:
(this causes other issues, let's tackle this in a different PR)

Reply:
Oooh interesting. I'm guessing VSCode isn't using the project version of typescript for checks like that for some reason